README.md
# A client to track events

## Usage

The client allows you to either track *impressions* which are nothing more but stupid, incrementing counters for a given event. Or you can track *uniques* which do increment their count only once for any given event and UUID combination in any resolution.

The client understands different resolutions to query events for. So e.g. you can query for the number of times the ``sign_up`` event happened last week, or last month or just today. All available resolutions are:

```ruby
[:year, :month, :week, :day, :hour]
```

Be aware of the fact, that impressions and uniques are two completely different topic. Even if you use the same event name they will not interfere or mix at all.

### Setup

```ruby
redis_url = "redis://user:password@yourhost:15540"
client = Tracker::Client.new(redis_url)
```

**Important**: All tracking operations are non-blocking by default! You can override that on a per call basis or set the default to be blocking. All examples below are using a blocking client!

```ruby
client = Tracker::Client.new(redis_url, blocking: true)
```

### Impressions

#### Tracking

```ruby
# Track an impression that happened right now
client.track_impression(:user_signup)

# Track an impression at a given time
when_it_happened = Time.new(2012, 10, 1)
client.track_impression(:user_signup, time: when_it_happened)

# Tracking an impression explicitly non-blocking / blocking
client.track_impression(:user_signup, blocking: false) # => returns Thread, so you can call #join on it
client.track_impression(:user_signup, blocking: true)
```

#### Querying

```ruby
# All user signups this week
client.query_impression(:user_signup, :week)

# All user signups last year
client.query_impression(:user_signup, :year, time: Time.new(Time.now.year - 1))
```

You can also query for multiple event/time combinations at once:

```ruby
query = [
  [:day],
  [:day, Time.now - (24 * 60 * 60)], # yesterday
  [:year, Time.new(Time.now.year - 1)], # last year
]
client.query_impression(:user_signup, query) # => [3, 5, 107]
```

### Uniqe events

E.g. unique page impression given a user. These do not add up but only count once for each user during any resolution.

#### Tracking

```ruby
# Track an event that happened right now
client.track_unique(:user_signup, uuid)

# Track an event at a given time
when_it_happened = Time.new(2012, 10, 1)
client.track_unique(:user_signup, uuid, time: when_it_happened)

# Tracking an unique explicitly non-blocking / blocking
client.track_unique(:user_signup, uuid, blocking: false) # => returns Thread, so you can call #join on it
client.track_unique(:user_signup, uuid, blocking: true)
```

#### Querying

```ruby
resolution = :month
when = Time.new(2012, 10)
result = client.query_unique(:user_signup, resolution, time: when)

# find out how many events occured during the timespan
result.length # => 1

# check if an event for a given UUID was tracked
result.include_uuid?(uuid) # => true / false

# query for signups this month (1st to last day in the last month)
client.query_unique(:user_signup, :month, time: Time.now)

# query for signups last year (1st of Jan to 31th of Dec)
last_year = Time.new(Time.now.year - 1)
client.query_unique(:user_signup, :year, time: last_year)
```

### Game specific tracking

#### Track an anonymous player

```ruby
client.game.track_player(game_uuid, "spiral-galaxy")

# Track at a specific point in time
last_year = Time.new(Time.now.year - 1)
client.game.track_player(game_uuid, "spiral-galaxy", time: last_year)
```

#### Track a logged in player

```ruby
client.game.track_logged_in_player(game_uuid, "spiral-galaxy")

# Track at a specific point in time
last_year = Time.new(Time.now.year - 1)
client.game.track_logged_in_player(game_uuid, "spiral-galaxy", time: last_year)
```

#### Retrieve insights on a game

```ruby
client.game.insights(game_uuid)

# get them for a specific point in time
last_year = Time.new(Time.now.year - 1)
client.game.insights(game_uuid, time: last_year)
```

The ``#insights`` method returns a hash with a block for each venue and one for overall numbers. Each of those blocks have two keys, ``:anonymous`` (for numbers on not logged in players) and ``:logged_in`` (for numbers of logged in players). That could look like this:

```ruby
{
  overall: {
    anonymous: {
      …
    },
    logged_in: {
      …
    }
  },
  facebook: {
    anonymous: {
      …
    },
    logged_in: {
      …
    }
  }
}
```

Each of the inner blocks now looks like this:

```ruby
{
  total: 13245,
  today: 45,
  week: 301,
  month: 720,
  year: 6329,
  rolling_30_days: [76,13,46,23,24,23,63,…]
}
```

with data on the total impressions, the impressions today so far, during the current week, the current month and the current year. In addition to that it gives you the impressions over the last 30 days day by day. The first element in that array is the number of impressions 31 days ago. The last element is the number of impressions yesterday.

CHANGELOG.md
# 0.0.6 / 2013-04-06

* Adds game specific tracking and insights

# 0.0.5 / 2013-04-06

* Adds multi impression querying
* Refines interface
* Makes tracking non-blocking by default

# 0.0.4 / 2013-04-05

* Adds support to track multiple events/impressions at once

# 0.0.3 / 2013-04-05

* Adds impression tracking

# 0.0.2 / 2013-04-05

* Adds test helpers

# 0.0.1 / 2013-04-05

* Initial release

spec/game_spec.rb
require_relative './spec_helper'

require 'timecop'

GAME_IMPRESSIONS_KEY_PREFIX = 'game-played'

class Array
  def sum
    inject(0) {|a,e| a + e}
  end
end

def game_impression_key(suffix, options)
  registered = options[:registered]
  "#{GAME_IMPRESSIONS_KEY_PREFIX}#{suffix}#{registered ? '-registered-player' : ''}"
end

def must_have_total_plays(count, options = {})
  @client.query_impression(game_impression_key(nil, options), :year).must_equal count
end

def must_have_game_plays(game, count, options = {})
  @client.query_impression(game_impression_key("-#{game}", options), :year).must_equal count
end

def must_have_venue_plays(venue, count, options = {})
  @client.query_impression(game_impression_key("-#{venue}", options), :year).must_equal count
end

def must_have_venue_game_plays(venue, game, count, options = {})
  @client.query_impression(game_impression_key("-#{venue}-#{game}", options), :year).must_equal count
end

def track_players(plays, game, venue, player = nil)
  now = Time.now
  day = 24 * 60 * 60

  i = plays.size
  plays.each do |count|
    i -= 1
    count.times do
      if player
        @client.game.track_logged_in_player(game, venue, player, time: now - (i * day))
      else
        @client.game.track_player(game, venue, time: now - (i * day))
      end
    end
  end
end

describe Tracking::Client::Game do
  before do
    @client = Tracking::Client.new('redis://localhost:6379/', blocking: true)

    Tracking::Client::TestHelpers.delete_all_tracked_events!(@client)

    @game = UUID.new.generate
    @game2 = UUID.new.generate

    Timecop.freeze(Time.new(2013, 7, 15, 1, 30))
  end

  after do
    Timecop.return
  end

  it "only tracks keys for unregistered players when tracking unregistered players" do
    @client.game.track_player(@game, "facebook")
    @client.game.track_player(@game, "facebook")
    @client.game.track_player(@game, "spiral-galaxy")

    @client.game.track_player(@game2, "facebook")

    must_have_total_plays(4)
    must_have_game_plays(@game, 3)
    must_have_game_plays(@game2, 1)
    must_have_venue_plays('facebook', 3)
    must_have_venue_plays('spiral-galaxy', 1)
    must_have_venue_game_plays('facebook', @game, 2)
    must_have_venue_game_plays('spiral-galaxy', @game, 1)
    must_have_venue_game_plays('facebook', @game2, 1)
    must_have_venue_game_plays('spiral-galaxy', @game2, 0)

    must_have_total_plays(0, registered: true)
    must_have_game_plays(@game, 0, registered: true)
    must_have_game_plays(@game2, 0, registered: true)
    must_have_venue_plays('facebook', 0, registered: true)
    must_have_venue_plays('spiral-galaxy', 0, registered: true)
    must_have_venue_game_plays('facebook', @game, 0, registered: true)
    must_have_venue_game_plays('spiral-galaxy', @game, 0, registered: true)
    must_have_venue_game_plays('facebook', @game2, 0, registered: true)
    must_have_venue_game_plays('spiral-galaxy', @game2, 0, registered: true)
  end

  it "only tracks keys for registered players when tracking registered players" do
    fake_player = UUID.new.generate
    @client.game.track_logged_in_player(@game, "facebook", fake_player)
    @client.game.track_logged_in_player(@game, "facebook", fake_player)
    @client.game.track_logged_in_player(@game, "spiral-galaxy", fake_player)

    @client.game.track_logged_in_player(@game2, "facebook", fake_player)

    must_have_total_plays(0)
    must_have_game_plays(@game, 0)
    must_have_game_plays(@game2, 0)
    must_have_venue_plays('facebook', 0)
    must_have_venue_plays('spiral-galaxy', 0)
    must_have_venue_game_plays('facebook', @game, 0)
    must_have_venue_game_plays('spiral-galaxy', @game, 0)
    must_have_venue_game_plays('facebook', @game2, 0)
    must_have_venue_game_plays('spiral-galaxy', @game2, 0)

    must_have_total_plays(4, registered: true)
    must_have_game_plays(@game, 3, registered: true)
    must_have_game_plays(@game2, 1, registered: true)
    must_have_venue_plays('facebook', 3, registered: true)
    must_have_venue_plays('spiral-galaxy', 1, registered: true)
    must_have_venue_game_plays('facebook', @game, 2, registered: true)
    must_have_venue_game_plays('spiral-galaxy', @game, 1, registered: true)
    must_have_venue_game_plays('facebook', @game2, 1, registered: true)
    must_have_venue_game_plays('spiral-galaxy', @game2, 0, registered: true)
  end

  it "can return basic insights on a game" do
    fake_player = UUID.new.generate

    plays_on_facebook = 40.times.map {rand(10)}
    plays_on_spiral_galaxy = 40.times.map {rand(10)}
    plays_on_embedded = 40.times.map {rand(10)}

    track_players(plays_on_facebook, @game, 'facebook')
    track_players(plays_on_spiral_galaxy, @game, 'spiral-galaxy')
    track_players(plays_on_embedded, @game, 'embedded')

    registered_plays_on_facebook = 40.times.map {rand(10)}
    registered_plays_on_spiral_galaxy = 40.times.map {rand(10)}
    registered_plays_on_embedded = 40.times.map {rand(10)}

    track_players(registered_plays_on_facebook, @game, 'facebook', fake_player)
    track_players(registered_plays_on_spiral_galaxy, @game, 'spiral-galaxy', fake_player)
    track_players(registered_plays_on_embedded, @game, 'embedded', fake_player)

    insights = @client.game.insights(@game)

    insights.keys.size.must_equal 4
    insights[:overall].must_equal(
      anonymous: {
        total: plays_on_facebook.sum + plays_on_spiral_galaxy.sum + plays_on_embedded.sum,
        today: plays_on_facebook.last + plays_on_spiral_galaxy.last + plays_on_embedded.last,
        week: plays_on_facebook[-2..-1].sum + plays_on_spiral_galaxy[-2..-1].sum + plays_on_embedded[-2..-1].sum,
        month: plays_on_facebook[-15..-1].sum + plays_on_spiral_galaxy[-15..-1].sum + plays_on_embedded[-15..-1].sum,
        year: plays_on_facebook.sum + plays_on_spiral_galaxy.sum + plays_on_embedded.sum,
        rolling_30_days: 30.times.map {|i| plays_on_facebook[-31 + i] + plays_on_spiral_galaxy[-31 + i] + plays_on_embedded[-31 + i]}
      },
      logged_in: {
        total: registered_plays_on_facebook.sum + registered_plays_on_spiral_galaxy.sum + registered_plays_on_embedded.sum,
        today: registered_plays_on_facebook.last + registered_plays_on_spiral_galaxy.last + registered_plays_on_embedded.last,
        week: registered_plays_on_facebook[-2..-1].sum + registered_plays_on_spiral_galaxy[-2..-1].sum + registered_plays_on_embedded[-2..-1].sum,
        month: registered_plays_on_facebook[-15..-1].sum + registered_plays_on_spiral_galaxy[-15..-1].sum + registered_plays_on_embedded[-15..-1].sum,
        year: registered_plays_on_facebook.sum + registered_plays_on_spiral_galaxy.sum + registered_plays_on_embedded.sum,
        rolling_30_days: 30.times.map {|i| registered_plays_on_facebook[-31 + i] + registered_plays_on_spiral_galaxy[-31 + i] + registered_plays_on_embedded[-31 + i]}
      }
    )
    insights[:"spiral-galaxy"].must_equal(
      anonymous: {
        total: plays_on_spiral_galaxy.sum,
        today: plays_on_spiral_galaxy.last,
        week: plays_on_spiral_galaxy[-2..-1].sum,
        month: plays_on_spiral_galaxy[-15..-1].sum,
        year: plays_on_spiral_galaxy.sum,
        rolling_30_days: 30.times.map {|i| plays_on_spiral_galaxy[-31 + i]}
      },
      logged_in: {
        total: registered_plays_on_spiral_galaxy.sum,
        today: registered_plays_on_spiral_galaxy.last,
        week: registered_plays_on_spiral_galaxy[-2..-1].sum,
        month: registered_plays_on_spiral_galaxy[-15..-1].sum,
        year: registered_plays_on_spiral_galaxy.sum,
        rolling_30_days: 30.times.map {|i| registered_plays_on_spiral_galaxy[-31 + i]}
      }
    )

    insights[:facebook].must_equal(
      anonymous: {
        total: plays_on_facebook.sum,
        today: plays_on_facebook.last,
        week: plays_on_facebook[-2..-1].sum,
        month: plays_on_facebook[-15..-1].sum,
        year: plays_on_facebook.sum,
        rolling_30_days: 30.times.map {|i| plays_on_facebook[-31 + i]}
      },
      logged_in: {
        total: registered_plays_on_facebook.sum,
        today: registered_plays_on_facebook.last,
        week: registered_plays_on_facebook[-2..-1].sum,
        month: registered_plays_on_facebook[-15..-1].sum,
        year: registered_plays_on_facebook.sum,
        rolling_30_days: 30.times.map {|i| registered_plays_on_facebook[-31 + i]}
      }
    )
    insights[:embedded].must_equal(
      anonymous: {
        total: plays_on_embedded.sum,
        today: plays_on_embedded.last,
        week: plays_on_embedded[-2..-1].sum,
        month: plays_on_embedded[-15..-1].sum,
        year: plays_on_embedded.sum,
        rolling_30_days: 30.times.map {|i| plays_on_embedded[-31 + i]}
      },
      logged_in: {
        total: registered_plays_on_embedded.sum,
        today: registered_plays_on_embedded.last,
        week: registered_plays_on_embedded[-2..-1].sum,
        month: registered_plays_on_embedded[-15..-1].sum,
        year: registered_plays_on_embedded.sum,
        rolling_30_days: 30.times.map {|i| registered_plays_on_embedded[-31 + i]}
      }
    )
  end
end

spec/client_spec.rb
require_relative './spec_helper'

require 'timecop'

describe Tracking::Client do
  before do
    @client = Tracking::Client.new('redis://localhost:6379/', blocking: true)

    Tracking::Client::TestHelpers.delete_all_tracked_events!(@client)

    @entity1 = UUID.new.generate
    @entity2 = UUID.new.generate
    @entity3 = UUID.new.generate

    Timecop.freeze(Time.new(2013, 7, 15, 1, 30))
  end

  after do
    Timecop.return
  end

  it "should be able to track and query unique events" do
    @client.track_unique(:login, @entity1)
    @client.track_unique(:login, @entity2)
    @client.track_unique(:login, @entity3, time: in_the_future(1 * month))

    @client.track_unique(:logout, @entity1, time: ago(1 * week))
    @client.track_unique(:logout, @entity3, time: in_the_future(1 * week))

    logins_this_year = @client.query_unique(:login, :year, time: Time.now)
    logins_this_year.length.must_equal 3
    logins_this_year.include_uuid?(@entity1).must_equal true
    logins_this_year.include_uuid?(@entity2).must_equal true
    logins_this_year.include_uuid?(@entity3).must_equal true

    logins_this_month = @client.query_unique(:login, :month)
    logins_this_month.length.must_equal 2
    logins_this_year.include_uuid?(@entity1).must_equal true
    logins_this_year.include_uuid?(@entity2).must_equal true

    logouts_two_weeks_ago = @client.query_unique(:logout, :week, time: ago(2 * week))
    logouts_two_weeks_ago.length.must_equal 0

    logouts_last_week = @client.query_unique(:logout, :week, time: ago(1 * week))
    logouts_last_week.length.must_equal 1
    logouts_last_week.include_uuid?(@entity1).must_equal true

    logouts_this_month = @client.query_unique(:logout, :month)
    logouts_this_month.length.must_equal(2)
    logouts_this_month.include_uuid?(@entity1).must_equal true
    logouts_this_month.include_uuid?(@entity3).must_equal true

    logouts_this_year = @client.query_unique(:logout, :year)
    logouts_this_year.length.must_equal(2)
    logouts_this_year.include_uuid?(@entity1).must_equal true
    logouts_this_year.include_uuid?(@entity3).must_equal true
  end

  it "can track impressions" do
    @client.track_impression(:login, time: ago(1 * year))

    @client.track_impression(:login, time: ago(2 * month))

    @client.track_impression(:login, time: ago(1 * week))
    @client.track_impression(:login, time: ago(1 * week))

    @client.track_impression(:login, time: ago(1 * day))

    @client.track_impression(:login)

    # watch out, it's a logout!
    @client.track_impression(:logout)

    @client.track_impression(:login, time: ago(2 * hour))
    @client.track_impression(:login, time: ago(1 * hour))

    @client.track_impression(:login, time: in_the_future(1 * week))

    # this year
    @client.query_impression(:login, :year).must_equal 8
    # last year
    @client.query_impression(:login, :year, time: ago(8 * month)).must_equal 1
    # this month
    @client.query_impression(:login, :month).must_equal 7
    # two month ago
    @client.query_impression(:login, :month, time: ago(2 * month)).must_equal 1
    # this week
    @client.query_impression(:login, :week).must_equal 4
    # last week
    @client.query_impression(:login, :week, time: ago(1 * week)).must_equal 2
    # today
    @client.query_impression(:login, :day).must_equal 2
    # yesterday
    @client.query_impression(:login, :day, time: ago(1 * day)).must_equal 2
    # next week
    @client.query_impression(:login, :week, time: in_the_future(1 * week)).must_equal 1

    # two years ago
    @client.query_impression(:login, :week, time: ago(2 * year)).must_equal 0
  end

  it "can track multiple impressions" do
    @client.track_impression([:login, :logout])
    @client.query_impression(:login, :year).must_equal 1
    @client.query_impression(:logout, :year).must_equal 1
  end

  it "can track total impressions" do
    @client.track_impression(:login, time: ago(3 * year))
    @client.track_impression(:login)
    @client.track_impression(:login, time: in_the_future(52 * year))

    @client.query_impression(:login, :total).must_equal 3
  end

  it "can query for impressions of an event during multiple times at once" do
    @client.track_impression(:login)
    @client.track_impression(:login, time: ago(1 * day))
    @client.track_impression(:login, time: ago(3 * day))
    @client.track_impression(:login, time: ago(1 * week))
    @client.track_impression(:login, time: ago(2 * month))

    query = [
      [:day],
      [:week, time: ago(1 * week)],
      [:month],
      [:month, time: ago(2 * month)],
      [:year],
      [:year, time: in_the_future(1 * year)]
    ]

    impressions = @client.query_impression(:login, query)
    impressions.must_equal [
      1,
      2,
      4,
      1,
      5,
      0
    ]
  end

  it "can track multiple events" do
    @client.track_unique([:login, :logout], @entity1)
    logins = @client.query_unique(:login, :year)
    logins.length.must_equal 1
    logins.include_uuid?(@entity1).must_equal true

    logouts = @client.query_unique(:logout, :year)
    logouts.length.must_equal 1
    logouts.include_uuid?(@entity1).must_equal true
  end

  it "can track impressions asynchronously" do
    lots_of_events = 10000.times.map {"event-#{rand(10000)}"}
    lots_of_events << 'login'
    tracking_thread = @client.track_impression(:login, blocking: false)

    @client.query_impression(:login, :year).must_equal 0
    tracking_thread.join
    @client.query_impression(:login, :year).must_equal 1
  end

  it "can track uniques asynchronously" do
    blocking_event = 'login'
    class << blocking_event
      attr_accessor :block_to_s

      alias raw_to_s to_s
      def to_s
        while @block_to_s; sleep 0.0001; end
        raw_to_s
      end
    end
    blocking_event.block_to_s = true

    tracking_thread = @client.track_unique(blocking_event, @entity1, blocking: false)

    @client.query_unique(:login, :year).length.must_equal 0
    blocking_event.block_to_s = false
    tracking_thread.join
    @client.query_unique(:login, :year).length.must_equal 1
  end
end

spec/spec_helper.rb
ENV['RACK_ENV'] ||= 'test'

Bundler.require

require 'minitest/autorun'

require 'tracking-client'
require 'tracking-client/test_helpers'

require_relative './spec_utility_methods'
include Tracking::Client::SpecUtilityMethods

lib/tracking-client.rb
require 'uuid'
require 'uuidtools'
require 'minuteman'

require "tracking-client/version"
require "tracking-client/error"
require "tracking-client/uuid_helpers"
require "tracking-client/query_result"
require "tracking-client/impression"
require "tracking-client/game"

module Tracking
  class Client
    SUPPORTED_RESOLUTIONS = %w{total year month week day hour}
    TRACKING_KEY_PREFIX = "qs_tracking_"

    attr_reader :game

    def initialize(redis_url, options = {})
      @minuteman = Minuteman.new(redis: {url: redis_url, driver: :hiredis}, silent: suppress_errors?, time_spans: SUPPORTED_RESOLUTIONS)
      @options = options
      @game = Game.new(self)
    end

    def track_unique(events, uuids, options = {})
      time = options[:time] || Time.now
      uuids = Array(uuids)
      action = Proc.new {
        Array(events).map do |event|
          @minuteman.track(event.to_s, uuids.map {|uuid| UUIDHelpers.uuid_to_int(uuid)}, time)
        end
      }

      run_action(action, options)
    end

    def query_unique(event, resolution, options = {})
      time = options[:time] || Time.now
      raise Error.new("Resolution `#{resolution}` not supported!") unless resolution_supported?(resolution)

      result = @minuteman.send(resolution, event.to_s, time)
      QueryResult.new(result)
    end

    def track_impression(events, options = {})
      time = options[:time] || Time.now

      action = Proc.new do
        Array(events).map do |event|
          Thread.new do
            Impression.new(event, time).keys.map do |key|
              Thread.new {redis.incr key}
            end.each(&:join)
          end
        end.each(&:join)
      end

      run_action(action, options)
    end

    def query_impression(event, resolutions, options = {})
      now = Time.now
      if resolutions.kind_of?(Array)
        keys = resolutions.map do |resolution, options|
          time = (options || {})[:time] || now
          Impression.new(event, time).key_for(resolution)
        end
        redis.mget(keys).map(&:to_i)
      else
        time = options[:time] || now
        redis.get(Impression.new(event, time).key_for(resolutions)).to_i
      end
    end

    private
    def suppress_errors?
      ENV['RACK_ENV'] == 'production'
    end

    def resolution_supported?(resolution)
      SUPPORTED_RESOLUTIONS.include?(resolution.to_s)
    end

    def redis
      @minuteman.redis
    end

    def run_action(action, options)
      if synchronous?(options)
        action.call
      else
        Thread.new(action) {|action| action.call}
      end
    end

    def synchronous?(options)
      options.has_key?(:blocking) ? options[:blocking] : @options[:blocking]
    end
  end
end

lib/tracking-client/impression.rb
module Tracking
  class Client
    class Impression
      def initialize(event, time)
        @event = event
        @time = time
      end

      def keys
        Client::SUPPORTED_RESOLUTIONS.map do |resolution|
          key_for(resolution)
        end
      end

      def key_for(resolution)
        "#{Client::TRACKING_KEY_PREFIX}#{@event}_#{time_key_for(resolution)}"
      end

      protected
      def time_key_for(resolution)
        send("time_key_for_#{resolution}")
      end

      def time_key_for_total
        "total"
      end

      def time_key_for_year
        "#{@time.year}"
      end

      def time_key_for_month
        "#{time_key_for_year}-#{@time.month}"
      end

      def time_key_for_week
        "#{time_key_for_month}-#{@time.strftime('%U')}"
      end

      def time_key_for_day
        "#{time_key_for_week}-#{@time.day}"
      end

      def time_key_for_hour
        "#{time_key_for_day}-#{@time.hour}"
      end
    end
  end
end

lib/tracking-client/game.rb
module Tracking
  class Client
    class Game
      TRACKING_KEY_PREFIX = "game-played"
      TRACKING_KEY_REGISTERED_SUFFIX = "registered-player"

      STARTING_YEAR = 2013
      VENUES = %w{facebook spiral-galaxy embedded}.map(&:to_sym)

      def initialize(client)
        @client = client
      end

      def track_logged_in_player(game_uuid, venue_name, player_uuid, options = {})
        @client.track_impression(registered_player_tracking_keys(game_uuid, venue_name), options)
      end

      def track_player(game_uuid, venue_name, options = {})
        @client.track_impression(tracking_keys(game_uuid, venue_name), options)
      end

      def insights(game_uuid, options = {})
        time = options[:time] || Time.now

        {
          overall: overall_insights(game_uuid, time)
        }.merge(venue_insights(game_uuid, time))
      end

      protected
      def tracking_keys(game_uuid, venue_name)
        keys = [
          "#{TRACKING_KEY_PREFIX}",
          "#{TRACKING_KEY_PREFIX}-#{game_uuid}",
          "#{TRACKING_KEY_PREFIX}-#{venue_name}",
          "#{TRACKING_KEY_PREFIX}-#{venue_name}-#{game_uuid}"
        ]
      end

      def registered_player_tracking_keys(game_uuid, venue_name)
        tracking_keys(game_uuid, venue_name).map {|k| "#{k}-#{TRACKING_KEY_REGISTERED_SUFFIX}"}
      end

      def overall_insights(game_uuid, time)
        insights_blocks_for("#{TRACKING_KEY_PREFIX}-#{game_uuid}", time)
      end

      def venue_insights(game_uuid, time)
        Hash[VENUES.map {|venue_name| [venue_name, insights_blocks_for("#{TRACKING_KEY_PREFIX}-#{venue_name}-#{game_uuid}", time)]}]
      end

      def insights_blocks_for(key, time)
        {
          anonymous: insights_block_for(key, time),
          logged_in: insights_block_for("#{key}-#{TRACKING_KEY_REGISTERED_SUFFIX}", time)
        }
      end

      def insights_block_for(key, time)
        day = 24 * 60 * 60

        query = []
        query << [:total, time: time]
        query << [:day, time: time]
        query << [:week, time: time]
        query << [:month, time: time]
        query << [:year, time: time]
        30.times.each do |i|
          query << [:day, time: time - (30 - i) * day]
        end
        impressions = @client.query_impression(key, query)

        {
          total: impressions.shift,
          today: impressions.shift,
          week: impressions.shift,
          month: impressions.shift,
          year: impressions.shift,
          rolling_30_days: impressions
        }
      end
    end
  end
end